linter/lints/cpcps/lint_crl_has_idp.go

```diff
@@ -69,91 +89,110 @@ func (l *crlHasIDP) Execute(c *x509.RevocationList) *lint.LintResult {
+	if !lints.ReadOptionalASN1BooleanWithTag(&idpv, &onlyContainsCACerts, onlyContainsCACertsTag, false) {
 		return &lint.LintResult{
 			Status:  lint.Error,
-			Details: "Failed to parse IDP URI",
+			Details: "Failed to read IDP onlyContainsCACerts",
 		}
 	}
 
-	if uri.Scheme != "http" {
+	if !idpv.Empty() {
 		return &lint.LintResult{
 			Status:  lint.Error,
-			Details: "IDP URI MUST use http scheme",
+			Details: "Unexpected IDP fields were found",
 		}
 	}
 
-	if !dpName.Empty() {
-		return &lint.LintResult{
-			Status:  lint.Warn,
-			Details: "IDP should contain only one distributionPoint",
+	if onlyContainsUserCerts {
+		if onlyContainsCACerts {
+			return &lint.LintResult{
+				Status:  lint.Error,
+				Details: "IDP should not have both onlyContainsUserCerts: TRUE and onlyContainsCACerts: TRUE",
+			}
 		}
-	}
 
-	// Ensure that OnlyContainsUserCerts is True. We have to read this boolean as
-	// a byte and ensure its value is 0xFF because cryptobyte.ReadASN1Boolean
-	// can't handle custom encoding rules like this field's [1] tag.
-	if !idpv.PeekASN1Tag(cryptobyte_asn1.Tag(1).ContextSpecific()) {
-		return &lint.LintResult{
-			Status:  lint.Warn,
-			Details: "IDP should contain onlyContainsUserCerts",
+		/*
+			RFC 5280 Section 4.2.1.13
+
+			DistributionPointName ::= CHOICE {
+				fullName                [0]     GeneralNames,
+				... }
+
+			RFC 5280 Appendix A.1, Page 128
+
+			GeneralName ::= CHOICE {
+				...
+			    uniformResourceIdentifier [6]  IA5String,
+				... }
+		*/
+		fullNameTag := cryptobyte_asn1.Tag(0).ContextSpecific().Constructed()
+		if !dpName.ReadASN1(&dpName, fullNameTag) {
+			return &lint.LintResult{
+				Status:  lint.Warn,
+				Details: "Failed to read IDP distributionPoint fullName",
+			}
 		}
-	}
 
-	onlyContainsUserCerts := make([]byte, 0)
-	if !idpv.ReadASN1Bytes(&onlyContainsUserCerts, cryptobyte_asn1.Tag(1).ContextSpecific()) {
-		return &lint.LintResult{
-			Status:  lint.Error,
-			Details: "Failed to read IDP onlyContainsUserCerts",
+		var uriBytes []byte
+		uriTag := cryptobyte_asn1.Tag(6).ContextSpecific()
+		if !dpName.ReadASN1Bytes(&uriBytes, uriTag) {
+			return &lint.LintResult{
+				Status:  lint.Warn,
+				Details: "Failed to read IDP URI",
+			}
 		}
-	}
 
-	if len(onlyContainsUserCerts) != 1 || onlyContainsUserCerts[0] != 0xFF {
-		return &lint.LintResult{
-			Status:  lint.Error,
-			Details: "IDP should set onlyContainsUserCerts: TRUE",
+		uri, err := url.Parse(string(uriBytes))
+		if err != nil {
+			return &lint.LintResult{
+				Status:  lint.Error,
+				Details: "Failed to parse IDP URI",
+			}
 		}
-	}
-
-	// Ensure that no other fields are set.
-	if !idpv.Empty() {
+		if uri.Scheme != "http" {
+			return &lint.LintResult{
+				Status:  lint.Error,
+				Details: "IDP URI MUST use http scheme",
+			}
+		}
+		if !dpName.Empty() {
+			return &lint.LintResult{
+				Status:  lint.Warn,
+				Details: "IDP should contain only one distributionPoint",
+			}
+		}
```

Comment:
Can you factor this out into a separate function, `parseSingleDistributionPointName`?